Pi+/Coxeter/Lehmer.agda

```diff
@@ -87,40 +87,234 @@ lemma-l++2++r a b (x :: l1) r1 (x₁ :: l2) r2 p with lemma-l++2++r a b l1 r1 l2
+      lemma3 = cut-prefix ((! lemma2 ∙ ! dfm) ∙ lemma1  ∙ ! (++-assoc (immersion cl) _ [ n ∸ rl ]))
+      lemma4 = cut-last ((! lemma2 ∙ ! dfm) ∙ lemma1  ∙ ! (++-assoc (immersion cl) _ [ n ∸ rl ]))
+      lemma5 = ++-assoc l (S (k + m) :: k + m :: m ↓ k) (S (k + m) :: rev r)
+  in  immersion-long-lemma (CanS cl {rl} (≤-down x)) k m l r ((ap (λ e -> immersion cl ++ e ↓ rl) (∸-up x) ∙ ! lemma4) ∙ lemma5)
+
+{-# NON_TERMINATING #-}
+final≅-Lehmer : {n : ℕ} -> (cl : Lehmer n) -> (m mf : List) -> (defm : m == (immersion {n} cl)) -> m ≅ mf -> ⊥
+final≅-Lehmer {O} CanZ nil mf defm p = empty-reduction p
+final≅-Lehmer {S O} (CanS CanZ {O} x) nil mf defm p = empty-reduction p
+final≅-Lehmer {S O} (CanS CanZ {S .0} (s≤s z≤n)) (x :: nil) mf defm p = one-reduction p
+final≅-Lehmer {S (S n)} (CanS (CanS cl {r₁} x₁) {r₂} x₂) m mf defm (cancel≅ {n₁} l r .m .mf defm₁ defmf) 
+  with (lemma-l++2++r n₁ n₁ ((immersion cl ++ S n ∸ r₁ ↓ r₁)) (S (S n) ∸ r₂ ↓ r₂) l r ((! defm) ∙ defm₁))
+... | R1 ((fst₁ , snd₁) , fst₂ , fst₃ , snd₂) = final≅-Lehmer {S n} (CanS cl {r₁} x₁) _ _ idp (cancel≅ _ _ _ _ fst₃ idp)
+... | R2 ((fst₁ , snd₁) , fst₂ , fst₃ , snd₂) = final≅-↓ (S (S n) ∸ r₂) r₂ _ (cancel≅ _ _ _ _ snd₂ idp)
+final≅-Lehmer {S (S n)} (CanS (CanS cl {r₁} x₁) {S r₂} x₂) m mf defm (cancel≅ {n₁} l r .m .mf defm₁ defmf) | R3 (fst₁ , snd₁) = 
+  >>-⊥ (S n) n₁ (≤-reflexive (! (plus-minus (≤-down2 x₂)) ∙ (cut-tail snd₁))) (immersion (CanS cl {r₁} x₁)) l nil (immersion->> (CanS cl {r₁} x₁)) fst₁
+final≅-Lehmer {S (S n)} (CanS (CanS cl {r₁} x₁) {r₂} x₂) m mf defm (swap≅ {n₁} {k₁} ps l r .m .mf defm₁ defmf) 
+  with (lemma-l++2++r n₁ k₁ ((immersion cl ++ S n ∸ r₁ ↓ r₁)) (S (S n) ∸ r₂ ↓ r₂) l r ((! defm) ∙ defm₁))
+... | R1 ((fst₁ , snd₁) , fst₂ , fst₃ , snd₂) = final≅-Lehmer {S n} (CanS cl {r₁} x₁) _ _ idp (swap≅ ps _ _ _ _ fst₃ idp)
+... | R2 ((fst₁ , snd₁) , fst₂ , fst₃ , snd₂) = final≅-↓ (S (S n) ∸ r₂) r₂ _ (swap≅ ps _ _ _ _ snd₂ idp)
+final≅-Lehmer {S (S n)} (CanS (CanS cl {r₁} x₁) {S r₂} x₂) m mf defm (swap≅ {n₁} {k₁} ps l r .m .mf defm₁ defmf) | R3 (fst₁ , snd₁) =
+  let lemma = 
+          ≤begin
+            S (S (S n))
+          ≡⟨ ! (ap (λ e -> S (S e)) (plus-minus (≤-down2 x₂))) ⟩ 
+            S (S (r₂ + S n ∸ r₂))
+          ≡⟨ ap (λ e -> S (S e)) (cut-tail snd₁) ⟩
+            S (S k₁)
+          ≤⟨ ps ⟩
+            n₁
+          ≤∎
+  in  >>-⊥ (S n) n₁ (≤-down (≤-down lemma)) (immersion (CanS cl {r₁} x₁)) l nil (immersion->> (CanS cl {r₁} x₁)) fst₁
+final≅-Lehmer {S (S n)} (CanS (CanS cl {r₁} x₁) {r₂} x₂) m mf defm (long≅ {n₁} k l r .m .mf defm₁ defmf) = 
+  let lemma = transport (λ e -> l ++ S (k + n₁) :: k + n₁ :: (n₁ ↓ k ++ S (k + n₁) :: e) == l ++ S (k + n₁) :: k + n₁ :: (n₁ ↓ k ++ S (k + n₁) :: rev (rev r))) (! rev-rev) idp
+  in  immersion-long-lemma (CanS (CanS cl x₁) x₂) k n₁ l (rev r) ( (! defm) ∙ defm₁ ∙ lemma)
```

Comment:
I don't know why Agda doesn't accept the termination condition - I am doing structural recursion on the first parameter.
@vikraman @JacquesCarette @sabry Any clue?

Reply:
Setting `--termination-depth=2` seems to work!

Reply:
Thanks! Fixed.